go/libkb/secret_store.go

```diff
@@ -246,3 +249,80 @@ func (s *SecretStoreLocked) GetUsersWithStoredSecrets(m MetaContext) ([]string,
 	}
 	return s.disk.GetUsersWithStoredSecrets(m)
 }
+
+func (s *SecretStoreLocked) PrimeSecretStores(mctx MetaContext) (err error) {
+	if s == nil || s.isNil() {
+		return errors.New("secret store is not available")
+	}
+	if s.disk != nil {
+		err = PrimeSecretStore(mctx, s.disk)
+		if err != nil {
+			return err
+		}
+	}
+	err = PrimeSecretStore(mctx, s.mem)
+	return err
+}
+
+// PrimeSecretStore runs a test with current platform's secret store, trying to
+// store, retrieve, and then delete a secret with an arbitrary name. This should
+// be done before provisioning or logging in
+func PrimeSecretStore(mctx MetaContext, ss SecretStoreAll) (err error) {
+	defer mctx.TraceTimed("PrimeSecretStore", func() error { return err })()
+
+	// Generate test username and test secret
+	testUsername, err := RandString("test_ss_", 5)
+	testUsername = strings.Replace(testUsername, "=", "_", -1)
+	if err != nil {
+		return err
+	}
+	randBytes, err := RandBytes(LKSecLen)
+	if err != nil {
+		return err
+	}
+	mctx.Debug("PrimeSecretStore: priming secret store with username %q and secret %v", testUsername, randBytes)
+	testNormUsername := NormalizedUsername(testUsername)
+	var secretF [LKSecLen]byte
+	copy(secretF[:], randBytes[:])
+	testSecret := LKSecFullSecret{f: &secretF}
+
+	shouldClear := true
+	defer func() {
+		if shouldClear {
+			err2 := ss.ClearSecret(mctx, testNormUsername)
+			mctx.Debug("PrimeSecretStore: Deferred clearing test secret store entry, returned: %v", err2)
+		}
+	}()
+
+	// Try to fetch first, we should get an error back.
+	_, err = ss.RetrieveSecret(mctx, testNormUsername)
+	if err == nil {
+		return errors.New("managed to retrieve secret before storing it")
+	} else if err != nil {
+		mctx.Debug("PrimeSecretStore: error when retrieving secret that wasn't stored yet: %q, as expected", err)
+	}
+
+	// Put secret in secret store through `SecretStore` interface.
+	err = ss.StoreSecret(mctx, testNormUsername, testSecret)
+	if err != nil {
+		return fmt.Errorf("error while storing secret: %s", err)
+	}
+
+	// Recreate test store with same username, try to retrieve secret.
+	retrSecret, err := ss.RetrieveSecret(mctx, testNormUsername)
+	if err != nil {
+		return fmt.Errorf("error while retrieving secret: %s", err)
+	}
+	mctx.Debug("PrimeSecretStore: Retrieved secret: %v", retrSecret.f)
+	if !retrSecret.Equal(testSecret) {
+		return errors.New("managed to retrieve test secret but it didn't match the stored one")
+	}
+
+	mctx.Debug("PrimeSecretStore: success - trying to clear test entry")
+	shouldClear = false
```

Comment:
why not have `shouldClear` determine if `err = err2` in the defer, then `ClearSecret` only has to get called once

Reply:
let me know if this is better now, thank you!